Enable backward/forward compatibility for TS runtime

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57778 Add enabling default instructions flag for mobile
* **#57498 Enable backward/forward compatibility for TS runtime**

We only ignore trailing unnecessary arguments because ignoring arguments between specified arguments makes this PR too complex. (e.g `aten::bar(specified, **unspecified**, specified, unspecified, unspecified)`. The reason is we need to mark all arguments keywords args because we have to tell the compiler which args are compiler. Once we do that, schema parser starts parsing the keyword args after args and complains that there are argument duplication (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/schema_matching.cpp#L432). This can/will be fixed in the next PR. 

Differential Revision: [D28162448](https://our.internmc.facebook.com/intern/diff/D28162448)

tugsbayasgalan committed:
Enable backward/forward compatibility for TS runtime

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Enable backward/forward compatibility for TS runtime"


We only ignore trailing unnecessary arguments because ignoring arguments between specified arguments makes this PR too complex. (e.g aten::bar(specified, **unspecified**, specified, unspecified, unspecified). The reason is we need to mark all arguments keywords args because we have to tell the compiler which args are compiler. Once we do that, schema parser starts parsing the keyword args after args and complains that there are argument duplication (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/schema_matching.cpp#L432). This can/will be fixed in the next PR. 

Differential Revision: [D28162448](https://our.internmc.facebook.com/intern/diff/D28162448)

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/frontend/schema_matching.h:
```diff
@@ -18,6 +18,10 @@ struct MatchedSchema {
   c10::OptNameList return_field_names;
 };
 
+TORCH_API int calculateNecessaryArgs(
+    const std::vector<Argument>& schema_args,
+    at::ArrayRef<Value*> actual_inputs);
+
```
I see that you moved this utility helper into `schema_matching` in order to make it accessible in multiple places. However, I don't seem the link between `calculateNecessayArgs` and `schema_matching`. Could you clarify? What do you think about putting it in a separate utility lib file?

tugsbayasgalan replied:
Sounds good. I just put there because they both reference Argument haha

gmagogsfm commented on torch/csrc/jit/frontend/schema_matching.cpp:
```diff
@@ -303,6 +303,36 @@ bool isBlockListedSchema(const FunctionSchema& schema) {
   return false;
 }
 
+int calculateNecessaryArgs(
```
```suggestion
int CalculateNecessaryArgs(
```
I think we should try to follow Google C++ style guide as much as possible:
https://google.github.io/styleguide/cppguide.html#Function_Names

gmagogsfm commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -1156,22 +1157,41 @@ struct PythonPrintImpl {
+          for (int index = 0; index < necessary_args; ++index) {
+            if (index > 0)
+              stmt << ", ";
+            auto v = useOf(node->inputs().at(index));
+            auto arg = schema.arguments().at(index);
             if (arg.kwarg_only()) {
               stmt << arg.name() << "=";
             }
-          } else {
-            // vararg functions like format can have extra arguments
-            AT_ASSERT(schema.is_vararg());
+            stmt << *v;
```
This `for` loop looks almost identical to the loop used in `is_vararg()` branch, other than the loops have different upper bound. Could we consider consolidating them into one?

tugsbayasgalan replied:
I thought it would help with readability. If we consolidate them together, I feel like it would be too complicated to understand.

gmagogsfm commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -1156,22 +1157,41 @@ struct PythonPrintImpl {
+          // details.
+          int necessary_args =
+              calculateNecessaryArgs(schema.arguments(), node->inputs());
+          for (int index = 0; index < necessary_args; ++index) {
```
Using `size_t` which is unsigned instead of `int` which is signed with a smaller range

gmagogsfm commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -1156,22 +1157,41 @@ struct PythonPrintImpl {
+              }
+            } else {
+              // vararg functions like format can have extra arguments
+              AT_ASSERT(schema.is_vararg());
```
This `assert` should never trigger since we are in false branch of `!schema.is_vararg()` when executing this line.

tugsbayasgalan replied:
yep will remove. This is from old code

gmagogsfm commented on test/test_jit.py:
```diff
@@ -4062,6 +4062,18 @@ def forward(self, x, w):
         with self.assertRaisesRegex(RuntimeError, 'test_jit.py\", line {}'.format(lineno + 3)):
             loaded(torch.rand(3, 4), torch.rand(30, 40))
 
+    def test_ignorable_args(self):
```
Generally we should avoid adding more tests to `test_jit.py`, which is growing too large to be manageable. Is there another test file that it fits in? If not, should we create one?

gmagogsfm commented on test/test_jit.py:
```diff
@@ -4062,6 +4062,18 @@ def forward(self, x, w):
             loaded(torch.rand(3, 4), torch.rand(30, 40))
 
+    def test_ignorable_args(self):
+        def fn():
+            val = torch.tensor([[1, 2, 3, 4, 9, 9], [1, 2, 3, 4, 9, 9]])
+            return val[2::, :1:]
```
Can we write IR directly and use "define" to make it into a function to avoid making this test case fragile?

Something like this:
https://github.com/pytorch/pytorch/blob/2a178d34cd2acd04c05050689a25c79e777c28c5/test/test_jit.py#L5060

tugsbayasgalan replied:
sounds good

gmagogsfm commented on test/test_jit.py:
```diff
@@ -4062,6 +4062,18 @@ def forward(self, x, w):
+            return val[2::, :1:]
+        scripted = torch.jit.script(fn)
+        src = str(scripted.code)
+        # we ignore trailing arguments after start=2 for dim 0
```
Optional:
It would be good if we include the schema of `slice` here to help with readability.

tugsbayasgalan replied:
sounds good

tugsbayasgalan committed:
Update on "Enable backward/forward compatibility for TS runtime"


We only ignore trailing unnecessary arguments because ignoring arguments between specified arguments makes this PR too complex. (e.g `aten::bar(specified, **unspecified**, specified, unspecified, unspecified)`. The reason is we need to mark all arguments keywords args because we have to tell the compiler which args are compiler. Once we do that, schema parser starts parsing the keyword args after args and complains that there are argument duplication (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/schema_matching.cpp#L432). This can/will be fixed in the next PR. 

Differential Revision: [D28162448](https://our.internmc.facebook.com/intern/diff/D28162448)

[ghstack-poisoned]

tugsbayasgalan commented on torch/csrc/jit/frontend/schema_matching.cpp:
```diff
@@ -303,6 +303,38 @@ bool isBlockListedSchema(const FunctionSchema& schema) {
+size_t CalculateNecessaryArgs(
+    const std::vector<Argument>& schema_args,
+    at::ArrayRef<Value*> actual_inputs) {
+  if (schema_args.size() < actual_inputs.size()) {
```
This is because it implies it is a vararg

tugsbayasgalan committed:
Update on "Enable backward/forward compatibility for TS runtime"


We only ignore trailing unnecessary arguments because ignoring arguments between specified arguments makes this PR too complex. (e.g `aten::bar(specified, **unspecified**, specified, unspecified, unspecified)`. The reason is we need to mark all arguments keywords args because we have to tell the compiler which args are compiler. Once we do that, schema parser starts parsing the keyword args after args and complains that there are argument duplication (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/schema_matching.cpp#L432). This can/will be fixed in the next PR. 

Differential Revision: [D28162448](https://our.internmc.facebook.com/intern/diff/D28162448)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Enable backward/forward compatibility for TS runtime"


We only ignore trailing unnecessary arguments because ignoring arguments between specified arguments makes this PR too complex. (e.g `aten::bar(specified, **unspecified**, specified, unspecified, unspecified)`. The reason is we need to mark all arguments keywords args because we have to tell the compiler which args are compiler. Once we do that, schema parser starts parsing the keyword args after args and complains that there are argument duplication (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/schema_matching.cpp#L432). This can/will be fixed in the next PR. 

Differential Revision: [D28162448](https://our.internmc.facebook.com/intern/diff/D28162448)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28162448).

gmagogsfm commented on test/jit/test_ignorable_args.py:
```diff
@@ -0,0 +1,47 @@
+    def test_slice_ignorable_args_for_slice(self):
+        graph_str = """graph():
+            %15 : int = prim::Constant[value=9223372036854775807]()
+            %13 : int = prim::Constant[value=0]() # test/test_jit.py:4068:19
```
Source location after # isn't necessary

tugsbayasgalan replied:
removed

gmagogsfm commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -1156,14 +1157,18 @@ struct PythonPrintImpl {
+            CalculateNecessaryArgs(schema.arguments(), node->inputs());
+        for (size_t index = 0; index < necessary_args; ++index) {
+          if (index > 0)
             stmt << ", ";
-          }
-          auto v = useOf(node->inputs().at(i));
+          auto v = useOf(node->inputs().at(index));
           // print the kwarg name if it is a kwarg only argument.
-          if (i < schema.arguments().size()) {
-            auto arg = schema.arguments().at(i);
+          if (index < schema.arguments().size()) {
+            auto arg = schema.arguments().at(index);
```
These changes are not necessary if we don't change the variable name from i to index

tugsbayasgalan replied:
done

gmagogsfm commented on torch/csrc/jit/frontend/schema_matching.h:
```diff
@@ -18,6 +18,10 @@ struct MatchedSchema {
   c10::OptNameList return_field_names;
 };
 
+TORCH_API size_t CalculateNecessaryArgs(
+    const std::vector<Argument>& schema_args,
+    at::ArrayRef<Value*> actual_inputs);
+
```
Are these still needed?

tugsbayasgalan replied:
removed

tugsbayasgalan committed:
Update on "Enable backward/forward compatibility for TS runtime"


We only ignore trailing unnecessary arguments because ignoring arguments between specified arguments makes this PR too complex. (e.g `aten::bar(specified, **unspecified**, specified, unspecified, unspecified)`. The reason is we need to mark all arguments keywords args because we have to tell the compiler which args are compiler. Once we do that, schema parser starts parsing the keyword args after args and complains that there are argument duplication (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/schema_matching.cpp#L432). This can/will be fixed in the next PR. 

Differential Revision: [D28162448](https://our.internmc.facebook.com/intern/diff/D28162448)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28162448).

tugsbayasgalan committed:
Update on "Enable backward/forward compatibility for TS runtime"


We only ignore trailing unnecessary arguments because ignoring arguments between specified arguments makes this PR too complex. (e.g `aten::bar(specified, **unspecified**, specified, unspecified, unspecified)`. The reason is we need to mark all arguments keywords args because we have to tell the compiler which args are compiler. Once we do that, schema parser starts parsing the keyword args after args and complains that there are argument duplication (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/schema_matching.cpp#L432). This can/will be fixed in the next PR. 

Differential Revision: [D28162448](https://our.internmc.facebook.com/intern/diff/D28162448)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28162448).

tugsbayasgalan committed:
Update on "Enable backward/forward compatibility for TS runtime"


We only ignore trailing unnecessary arguments because ignoring arguments between specified arguments makes this PR too complex. (e.g `aten::bar(specified, **unspecified**, specified, unspecified, unspecified)`. The reason is we need to mark all arguments keywords args because we have to tell the compiler which args are compiler. Once we do that, schema parser starts parsing the keyword args after args and complains that there are argument duplication (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/schema_matching.cpp#L432). This can/will be fixed in the next PR. 

Differential Revision: [D28162448](https://our.internmc.facebook.com/intern/diff/D28162448)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28162448).

tugsbayasgalan committed:
Update on "Enable backward/forward compatibility for TS runtime"


We only ignore trailing unnecessary arguments because ignoring arguments between specified arguments makes this PR too complex. (e.g `aten::bar(specified, **unspecified**, specified, unspecified, unspecified)`. The reason is we need to mark all arguments keywords args because we have to tell the compiler which args are compiler. Once we do that, schema parser starts parsing the keyword args after args and complains that there are argument duplication (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/schema_matching.cpp#L432). This can/will be fixed in the next PR. 

Differential Revision: [D28162448](https://our.internmc.facebook.com/intern/diff/D28162448)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28162448).

tugsbayasgalan committed:
Update on "Enable backward/forward compatibility for TS runtime"


We only ignore trailing unnecessary arguments because ignoring arguments between specified arguments makes this PR too complex. (e.g `aten::bar(specified, **unspecified**, specified, unspecified, unspecified)`. The reason is we need to mark all arguments keywords args because we have to tell the compiler which args are compiler. Once we do that, schema parser starts parsing the keyword args after args and complains that there are argument duplication (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/schema_matching.cpp#L432). This can/will be fixed in the next PR. 

Differential Revision: [D28162448](https://our.internmc.facebook.com/intern/diff/D28162448)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28162448).

tugsbayasgalan committed:
Update on "Enable backward/forward compatibility for TS runtime"


We only ignore trailing unnecessary arguments because ignoring arguments between specified arguments makes this PR too complex. (e.g `aten::bar(specified, **unspecified**, specified, unspecified, unspecified)`. The reason is we need to mark all arguments keywords args because we have to tell the compiler which args are compiler. Once we do that, schema parser starts parsing the keyword args after args and complains that there are argument duplication (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/schema_matching.cpp#L432). This can/will be fixed in the next PR. 

Differential Revision: [D28162448](https://our.internmc.facebook.com/intern/diff/D28162448)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28162448).

tugsbayasgalan committed:
Update on "Enable backward/forward compatibility for TS runtime"


We only ignore trailing unnecessary arguments because ignoring arguments between specified arguments makes this PR too complex. (e.g `aten::bar(specified, **unspecified**, specified, unspecified, unspecified)`. The reason is we need to mark all arguments keywords args because we have to tell the compiler which args are compiler. Once we do that, schema parser starts parsing the keyword args after args and complains that there are argument duplication (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/schema_matching.cpp#L432). This can/will be fixed in the next PR. 

Differential Revision: [D28162448](https://our.internmc.facebook.com/intern/diff/D28162448)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28162448).

tugsbayasgalan committed:
Update on "Enable backward/forward compatibility for TS runtime"


We only ignore trailing unnecessary arguments because ignoring arguments between specified arguments makes this PR too complex. (e.g `aten::bar(specified, **unspecified**, specified, unspecified, unspecified)`. The reason is we need to mark all arguments keywords args because we have to tell the compiler which args are compiler. Once we do that, schema parser starts parsing the keyword args after args and complains that there are argument duplication (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/schema_matching.cpp#L432). This can/will be fixed in the next PR. 

Differential Revision: [D28162448](https://our.internmc.facebook.com/intern/diff/D28162448)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28162448).

tugsbayasgalan committed:
Update on "Enable backward/forward compatibility for TS runtime"


We only ignore trailing unnecessary arguments because ignoring arguments between specified arguments makes this PR too complex. (e.g `aten::bar(specified, **unspecified**, specified, unspecified, unspecified)`. The reason is we need to mark all arguments keywords args because we have to tell the compiler which args are compiler. Once we do that, schema parser starts parsing the keyword args after args and complains that there are argument duplication (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/schema_matching.cpp#L432). This can/will be fixed in the next PR. 

Differential Revision: [D28162448](https://our.internmc.facebook.com/intern/diff/D28162448)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28162448).
